core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushPartialAggregationThroughJoin.java

```diff
@@ -109,20 +178,75 @@ private static boolean allAggregationsOn(Map<Symbol, Aggregation> aggregations,
-        AggregationNode pushedAggregation = replaceAggregationSource(node, child.getLeft(), groupingSet);
-        return pushPartialToJoin(node, child, pushedAggregation, child.getRight(), context);
+        return getPushedAggregation(node, child, child.getRight(), context)
+                .map(pushedAggregation -> replaceJoin(node, pushedAggregation, child, child.getLeft(), pushedAggregation, context));
     }
 
-    private PlanNode pushPartialToRightChild(AggregationNode node, JoinNode child, Context context)
+    private Optional<AggregationNode> getPushedAggregation(AggregationNode node, JoinNode child, PlanNode joinSource, Context context)
     {
-        Set<Symbol> joinRightChildSymbols = ImmutableSet.copyOf(child.getRight().getOutputSymbols());
-        List<Symbol> groupingSet = getPushedDownGroupingSet(node, joinRightChildSymbols, intersection(getJoinRequiredSymbols(child), joinRightChildSymbols));
-        AggregationNode pushedAggregation = replaceAggregationSource(node, child.getRight(), groupingSet);
-        return pushPartialToJoin(node, child, child.getLeft(), pushedAggregation, context);
+        Set<Symbol> joinSourceSymbols = ImmutableSet.copyOf(joinSource.getOutputSymbols());
+        List<Symbol> groupingSet = getPushedDownGroupingSet(node, joinSourceSymbols, intersection(getJoinRequiredSymbols(child), joinSourceSymbols));
+        AggregationNode pushedAggregation = replaceAggregationSource(node, joinSource, groupingSet);
+        if (skipPartialAggregationPushdown(child, node, pushedAggregation, context)) {
+            return Optional.empty();
+        }
+        return Optional.of(pushedAggregation);
+    }
+
+    private boolean skipPartialAggregationPushdown(
+            JoinNode join,
+            AggregationNode originalAggregation,
+            AggregationNode pushedAggregation,
+            Context context)
+    {
+        // Only push aggregation down if pushed aggregation consumes similar number of input rows (e.g. join is not expanding).
+        // Otherwise, there is a possibility that aggregation above join could be more efficient in reducing total number of rows.
+        PlanNodeStatsEstimate sourceStats = context.getStatsProvider().getStats(pushedAggregation.getSource());
+        double sourceRowCount = sourceStats.getOutputRowCount();
+        double joinRowCount = context.getStatsProvider().getStats(join).getOutputRowCount();
+        // Pushing aggregation though filtering join could mean more work for partial aggregation. However,
+        // we allow pushing partial aggregations through filtering join because:
+        // 1. dynamic filtering should filter unmatched rows at source
+        // 2. partial aggregation will adaptively switch off when it's not reducing input rows
+        // 3. join operator is not particularly effective at filtering rows
+        if (isNaN(sourceRowCount) || isNaN(joinRowCount) || joinRowCount > 1.1 * sourceRowCount) {
```

Comment:
Why "1.1"?

Reply:
We allow for a bit of flexibility here (e.g. in case of small estimation inaccuracies)

Reply:
Thanks. Add a comment to explain that reasoning.

Reply:
done